src/lavinmq/mqtt/client.cr

```diff
@@ -0,0 +1,232 @@
+require "openssl"
+require "socket"
+require "../client"
+require "../error"
+require "./session"
+
+module LavinMQ
+  module MQTT
+    class Client < LavinMQ::Client
+      include Stats
+      include SortableJSON
+
+      getter vhost, channels, log, name, user, client_id, socket
+      @channels = Hash(UInt16, Client::Channel).new
+      @session : MQTT::Session?
+      rate_stats({"send_oct", "recv_oct"})
+      Log = ::Log.for "mqtt.client"
+
+      def initialize(@socket : ::IO,
+                     @connection_info : ConnectionInfo,
+                     @user : User,
+                     @vhost : VHost,
+                     @broker : MQTT::Broker,
+                     @client_id : String,
+                     @clean_session = false,
+                     @will : MQTT::Will? = nil)
+        @io = MQTT::IO.new(@socket)
+        @lock = Mutex.new
+        @remote_address = @connection_info.src
+        @local_address = @connection_info.dst
+        @name = "#{@remote_address} -> #{@local_address}"
+        @metadata = ::Log::Metadata.new(nil, {vhost: @broker.vhost.name, address: @remote_address.to_s, client_id: client_id})
+        @log = Logger.new(Log, @metadata)
+        @broker.vhost.add_connection(self)
+        @log.info { "Connection established for user=#{@user.name}" }
+        spawn read_loop
+      end
+
+      def client_name
+        "mqtt-client"
+      end
+
+      private def read_loop
+        loop do
+          @log.trace { "waiting for packet" }
+          packet = read_and_handle_packet
+          # The disconnect packet has been handled and the socket has been closed.
+          # If we dont breakt the loop here we'll get a IO/Error on next read.
+          break if packet.is_a?(MQTT::Disconnect)
+        end
+      rescue ex : ::MQTT::Protocol::Error::PacketDecode
+        @log.warn(exception: ex) { "Packet decode error" }
+        publish_will if @will
+        @socket.close
+      rescue ex : MQTT::Error::Connect
+        @log.warn { "Connect error: #{ex.message}" }
+      rescue ex : ::IO::Error
+        @log.warn(exception: ex) { "Read Loop error" }
+        publish_will if @will
+      rescue ex
+        publish_will if @will
+        raise ex
```

Comment:
I wonder if we really should reraise here? Who will rescue it? Will it crash lavinmq? I wonder if we ever ends up here?
Maybe it should be logged with error and not reraised?

Reply:
yes, i dont know if i like raising either, i think logging a warning is good.

Reply:
idk if we can get there.. but if something strange happens, its nice to rescue it